pygfx/renderers/wgpu/_renderer.py

```diff
@@ -465,6 +465,20 @@ def render(
                 "The viewport rect must be None or 4 elements (x, y, w, h)."
             )
 
+        # Allow objects to prepare just in time. When doing multiple
+        # render calls, we don't want to spam. The clear_color flag is
+        # a good indicator to detect the first render call.
+        if clear_color:
+            ev = WindowEvent(
+                "before_render",
+                target=None,
+                root=self,
+                width=logical_size[0],
+                height=logical_size[1],
+                pixel_ratio=self.pixel_ratio,
+            )
+            self.dispatch_event(ev)
```

Comment:
I'm still not convinced of the benefit of this event over just making users call `controller.update()` in the `animate` function, [as done here](https://github.com/mrdoob/three.js/blob/47b28bc564b438bf2b80d6e5baf90235292fcbd7/examples/misc_controls_orbit.html#L130) ([example](https://threejs.org/examples/#misc_controls_orbit)).

Reply:
Well, is that true though? This is the first "renderer event" that isn't an "input event".

Reply:
Fine, let's roll with this, and we'll see how well it works!